File: bms_media/audio.py
import os
import subprocess
import time
import multiprocessing
from typing import List, Optional, Tuple


"""
Audio
"""


class AudioPreset:
    def __init__(
        self, exec: str, output_format: str, arg: Optional[str] = None
    ) -> None:
        self.exec = exec
        self.output_format = output_format
        self.arg = arg

    def __str__(self) -> str:
        return f"AudioPreset {{ exec: {self.exec}, output_format: {self.output_format} arg: {self.arg} }}"

    def __repr__(self) -> str:
        return self.__str__()


AUDIO_PRESET_OGG_Q10 = AudioPreset("oggenc", "ogg", "-q10")
AUDIO_PRESET_OGG_FFMPEG = AudioPreset("ffmpeg", "ogg", "")

AUDIO_PRESET_WAV_FFMPEG = AudioPreset("ffmpeg", "wav", None)
AUDIO_PRESET_WAV_FROM_FLAC = AudioPreset(
    "flac", "wav", "-d --keep-foreign-metadata-if-present -f"
)

AUDIO_PRESET_FLAC = AudioPreset(
    "flac", "flac", "--keep-foreign-metadata-if-present --best -f"
)
AUDIO_PRESET_FLAC_FFMPEG = AudioPreset("ffmpeg", "flac", "")


def _get_audio_precess_cmd(
    file_path: str,
    output_file_path: str,
    preset: AudioPreset,
) -> str:
    # Execute
    arg = preset.arg if preset.arg is not None else ""
    if preset.exec == "ffmpeg":
        return f'ffmpeg -hide_banner -loglevel panic -i "{file_path}" -f {preset.output_format} -map_metadata 0 {arg} "{output_file_path}"'
    elif preset.exec == "oggenc":
        return f'oggenc {arg} "{file_path}" -o "{output_file_path}"'
    elif preset.exec == "flac":
        return f'flac {arg} "{file_path}" -o "{output_file_path}"'
    else:
        return ""


def transfer_audio_by_format_in_dir(
    dir: str,
    input_exts: List[str],
    presets: List[AudioPreset],
    remove_origin_file: bool = True,
    remove_existing_target_file: bool = True,
) -> bool:
    """
    Example:
    wav flac flac
    wav ogg ogg -ab 320k
    """

    def check_input_file(
        dir: str, file_name: str, input_exts: List[str]
    ) -> Optional[str]:
        file_path = f"{dir}/{file_name}"
        if not os.path.isfile(file_path):
            return None
        # Check ext
        ext_found: Optional[str] = None
        for ext in input_exts:
            if file_path.lower().endswith("." + ext):
                ext_found = ext
        if ext_found is None:
            return None
        return os.path.join(dir, file_name)

    def spawn_parse_audio_process(
        file_path: str, preset_index: int, preset: AudioPreset
    ) -> Tuple[Tuple[str, int], Optional[subprocess.Popen]]:
        # New cmd
        output_file_path = (
            file_path[: -len(file_path.rsplit(".")[-1])] + preset.output_format
        )
        # Target File exists?
        if os.path.isfile(output_file_path):
            if (
                os.path.getsize(output_file_path) > 0
                and not remove_existing_target_file
            ):
                print(f" - File {output_file_path} exists! Skipping...")
                return (file_path, preset_index), None
            else:
                print(f" - Remove existing file: {output_file_path}")
                os.remove(output_file_path)
        # Run cmd
        cmd = _get_audio_precess_cmd(
            file_path,
            output_file_path,
            preset,
        )
        if len(cmd) == 0:
            return (file_path, preset_index), None
        process = subprocess.Popen(
            cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )
        return (file_path, preset_index), process

    has_error = False
    err_file_path = ""
    err_stdout = b""
    err_stderr = b""

    # 创建线程池
    hdd = not dir.upper().startswith("C:")
    max_workers = (
        min(multiprocessing.cpu_count(), 24) if hdd else multiprocessing.cpu_count()
    )

    # Submit
    processes: List[Tuple[Tuple[str, int], Optional[subprocess.Popen]]] = []
    task_args: List[Tuple[str, int, AudioPreset]] = [
        (os.path.join(dir, file_name), 0, presets[0])
        for file_name in os.listdir(dir)
        if check_input_file(dir, file_name, input_exts) is not None
    ]

    # Count
    file_count = len(task_args)
    fallback_files: List[Tuple[str, int]] = []

    for task_arg in task_args:
        if len(processes) >= max_workers:
            break
        processes.append(spawn_parse_audio_process(*task_arg))
    task_args = task_args[len(processes) :]

    # 等待所有任务完成
    while len(processes) > 0:
        new_processes: List[Tuple[Tuple[str, int], Optional[subprocess.Popen]]] = []

        # 检查进程状态
        switch_next_list: List[Tuple[str, int]] = []
        for process in processes:
            (file_path, preset_index), process = process
            # Switch Next?
            switch_next = False
            if process is None:
                # Empty process
                switch_next = True
            else:
                process_returncode = process.poll()
                if process_returncode is None:
                    # Running
                    new_processes.append(((file_path, preset_index), process))
                elif process_returncode == 0:
                    # Succcess
                    if remove_origin_file and os.path.isfile(file_path):
                        os.remove(file_path)
                else:
                    # Failed
                    switch_next = True
                    err_stdout, err_stderr = process.communicate()

            if switch_next:
                switch_next_list.append((file_path, preset_index))

        # 切换下个预设
        for file_path, preset_index in switch_next_list:
            new_preset_index = preset_index + 1
            if new_preset_index not in range(0, len(presets)):
                # Last, Return
                has_error = True
                continue
            # Count
            fallback_files.append((file_path, new_preset_index))
            # Try Next
            task_args.append((file_path, new_preset_index, presets[new_preset_index]))

        # 启动新进程
        running_count_delta = 0
        for task_arg in task_args:
            if len(new_processes) >= max_workers:
                break
            new_processes.append(spawn_parse_audio_process(*task_arg))
            running_count_delta += 1
        task_args = task_args[running_count_delta:]

        processes = new_processes

        # 休眠一阵子
        time.sleep(0.001)

    if has_error:
        print("Has Error!")
        print("- Err file_path: ", err_file_path)
        print("- Err stdout: ", err_stdout)
        print("- Err stderr: ", err_stderr)

    if file_count > 0:
        print(f" -v- Parsed {file_count} file(s).")
    if len(fallback_files) > 0:
        print(f" x_x Fallback: {fallback_files}.")

    return not has_error

File: be-music-cabinet/src/media/audio.rs
use std::{
    cell::LazyCell,
    collections::HashMap,
    ffi::OsStr,
    path::Path,
    process::{ExitStatus, Output},
    sync::Arc,
};

use smol::{
    fs::{self, remove_file},
    io,
    lock::Semaphore,
    process::Command,
    stream::StreamExt,
};

/// Audio processing preset configuration
#[derive(Debug, Clone)]
pub struct AudioPreset {
    /// Executor name (e.g., "ffmpeg", "oggenc")
    executor: String,
    /// Output format (e.g., "ogg", "flac")
    output_format: String,
    /// Additional arguments (can be empty)
    arguments: Option<String>,
}

impl AudioPreset {
    /// Create new audio preset
    fn new(executor: &str, output_format: &str, arguments: Option<&str>) -> Self {
        Self {
            executor: executor.to_string(),
            output_format: output_format.to_string(),
            arguments: arguments.map(|s| s.to_string()),
        }
    }
}

#[allow(clippy::declare_interior_mutable_const)]
pub const AUDIO_PRESETS: LazyCell<HashMap<&'static str, AudioPreset>> = LazyCell::new(|| {
    let mut map = HashMap::new();
    map.insert(
        "FLAC",
        AudioPreset::new(
            "flac",
            "flac",
            Some("--keep-foreign-metadata-if-present --best -f"),
        ),
    );
    map.insert("FLAC_FFMPEG", AudioPreset::new("ffmpeg", "flac", None));
    map.insert(
        "WAV_FROM_FLAC",
        AudioPreset::new(
            "flac",
            "wav",
            Some("-d --keep-foreign-metadata-if-present -f"),
        ),
    );
    map.insert("WAV_FFMPEG", AudioPreset::new("ffmpeg", "wav", None));
    map.insert("OGG_Q10", AudioPreset::new("oggenc", "ogg", Some("-q10")));
    map.insert("OGG_FFMPEG", AudioPreset::new("ffmpeg", "ogg", None));
    map
});

/// Get command string for processing audio files
///
/// # Parameters
/// - `input_path`: input file path
/// - `output_path`: output file path
/// - `preset`: audio preset to use
///
/// # Returns
/// Generated command line string
fn get_audio_command(
    input_path: &Path,
    output_path: &Path,
    preset: &AudioPreset,
) -> Option<String> {
    match preset.executor.as_str() {
        "ffmpeg" => {
            let args = preset.arguments.as_deref().unwrap_or("");
            Some(format!(
                "ffmpeg -hide_banner -loglevel panic -i \"{}\" -f {} -map_metadata 0 {} \"{}\"",
                input_path.display(),
                preset.output_format,
                args,
                output_path.display()
            ))
        }
        "oggenc" => {
            let args = preset.arguments.as_deref().unwrap_or("");
            Some(format!(
                "oggenc {} \"{}\" -o \"{}\"",
                args,
                input_path.display(),
                output_path.display()
            ))
        }
        "flac" => {
            let args = preset.arguments.as_deref().unwrap_or("");
            Some(format!(
                "flac {} \"{}\" -o \"{}\"",
                args,
                input_path.display(),
                output_path.display()
            ))
        }
        _ => None,
    }
}

/// Convert audio files in specified directory
///
/// # Parameters
/// - `dir_path`: target directory path
/// - `input_extensions`: list of input file extensions to process
/// - `presets`: list of presets to try in order
/// - `remove_on_success`: remove original file after successful conversion
/// - `remove_on_fail`: remove original file after all attempts fail
/// - `remove_existing`: whether to overwrite existing output files
///
/// # Returns
/// Whether the conversion operation was completely successful
async fn transfer_audio_in_directory(
    dir_path: &Path,
    input_extensions: &[&str],
    presets: &[AudioPreset],
    remove_on_success: bool,
    remove_on_fail: bool,
    remove_existing: bool,
) -> io::Result<bool> {
    let mut tasks = Vec::new();
    let mut total_files = 0;
    let mut fallback_files = Vec::new();
    let mut has_error = false;
    let mut last_error = None;

    // Determine concurrency limit based on whether it's on C drive (use more threads for non-C drives)
    let is_hdd = !dir_path.starts_with("C:");
    let max_workers = if is_hdd {
        num_cpus::get().min(24)
    } else {
        num_cpus::get()
    };
    let semaphore = Arc::new(Semaphore::new(max_workers));

    // Collect files to process
    let mut entries = fs::read_dir(dir_path).await?;
    while let Some(entry) = entries.next().await {
        let entry = entry?;
        let path = entry.path();
        if !path.is_file() {
            continue;
        }

        if let Some(ext) = path.extension().and_then(OsStr::to_str)
            && input_extensions.iter().any(|e| e.eq_ignore_ascii_case(ext))
        {
            total_files += 1;
            tasks.push(path.clone());
        }
    }

    if total_files > 0 {
        log::info!(
            "Entering dir: {}, Input extensions: {:?}",
            dir_path.display(),
            input_extensions
        );
        log::info!("Using presets: {presets:?}");
    }

    // Process each file
    for file_path in tasks {
        let mut current_preset_index = 0;
        let mut success = false;

        while current_preset_index < presets.len() {
            let preset = &presets[current_preset_index];
            let output_path = file_path.with_extension(&preset.output_format);

            // Check if output file exists
            if output_path.exists() {
                if remove_existing {
                    if let Ok(metadata) = fs::metadata(&output_path).await
                        && metadata.len() > 0
                    {
                        log::info!("Removing existing file: {}", output_path.display());
                        let _ = remove_file(&output_path).await;
                    }
                } else {
                    log::info!("Skipping existing file: {}", output_path.display());
                    current_preset_index += 1;
                    continue;
                }
            }

            // Get and execute command
            if let Some(cmd) = get_audio_command(&file_path, &output_path, preset) {
                let permit = semaphore.clone().acquire_arc().await;

                // Execute conversion command
                #[cfg(target_family = "windows")]
                let program = "powershell";
                #[cfg(not(target_family = "windows"))]
                let program = "sh";
                let output = Command::new(program).arg("-c").arg(&cmd).output().await;

                drop(permit); // Release semaphore

                match output {
                    Ok(output) if output.status.success() => {
                        // Conversion successful
                        if remove_on_success && let Err(e) = remove_file(&file_path).await {
                            eprintln!(
                                "Error deleting original file: {} - {}",
                                file_path.display(),
                                e
                            );
                        }
                        success = true;
                        break;
                    }
                    Ok(output) => {
                        // Conversion failed
                        log::info!(
                            "Preset failed [{}]: {} -> {}",
                            preset.executor,
                            file_path.display(),
                            output_path.display()
                        );
                        last_error = Some((file_path.clone(), output));
                    }
                    Err(e) => {
                        // Command execution error
                        eprintln!("Command execution error: {e}");
                        last_error = Some((
                            file_path.clone(),
                            Output {
                                status: ExitStatus::default(),
                                stdout: Vec::new(),
                                stderr: e.to_string().into_bytes(),
                            },
                        ));
                    }
                }
            }

            current_preset_index += 1;
        }

        if !success {
            has_error = true;
            fallback_files.push(file_path.file_name().unwrap().to_string_lossy().to_string());

            // Remove original file after all preset attempts fail
            if remove_on_fail && let Err(e) = remove_file(&file_path).await {
                eprintln!(
                    "Error deleting failed file: {} - {}",
                    file_path.display(),
                    e
                );
            }
        }
    }

    // Output processing results
    if total_files > 0 {
        log::info!("Processed {} files in {}", total_files, dir_path.display());
    }
    if !fallback_files.is_empty() {
        log::info!(
            "{} files failed all presets: {:?}",
            fallback_files.len(),
            fallback_files
        );
    }
    if has_error {
        if let Some((err_path, output)) = last_error {
            eprintln!("Last error on file: {}", err_path.display());
            eprintln!("stdout: {}", String::from_utf8_lossy(&output.stdout));
            eprintln!("stderr: {}", String::from_utf8_lossy(&output.stderr));
        }
        if remove_on_fail {
            log::info!("Original files for failed conversions were removed");
        }
    }

    Ok(!has_error)
}

/// Process all BMS folders under root directory
///
/// # Parameters
/// - `root_dir`: root directory path
/// - `input_extensions`: list of input file extensions
/// - `preset_names`: list of preset names to use
/// - `remove_on_success`: remove original file on success
/// - `remove_on_fail`: remove original file on failure
/// - `skip_on_fail`: skip subsequent processing on error
pub async fn process_bms_folders(
    root_dir: &Path,
    input_extensions: &[&str],
    preset_names: &[&str],
    remove_on_success: bool,
    remove_on_fail: bool,
    skip_on_fail: bool,
) -> io::Result<()> {
    // Parse preset names into preset objects
    let presets: Vec<AudioPreset> = preset_names
        .iter()
        .filter_map(|name| {
            let binding = AUDIO_PRESETS;
            let preset = binding.get(name);
            preset.cloned()
        })
        .collect();

    if presets.is_empty() {
        io::Error::other("No valid presets provided");
    }

    // Iterate through all subdirectories under root directory
    let mut entries = fs::read_dir(root_dir).await?;
    while let Some(entry) = entries.next().await {
        let entry = entry?;
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }

        log::info!("Processing directory: {}", dir_path.display());
        match transfer_audio_in_directory(
            &dir_path,
            input_extensions,
            &presets,
            remove_on_success,
            remove_on_fail,
            true, // Always overwrite existing files
        )
        .await
        {
            Ok(true) => log::info!("Successfully processed {}", dir_path.display()),
            Ok(false) => {
                eprintln!("Errors occurred in {}", dir_path.display());
                if skip_on_fail {
                    eprintln!("Skipping remaining folders due to error");
                    break;
                }
            }
            Err(e) => {
                eprintln!("Error processing {}: {}", dir_path.display(), e);
                if skip_on_fail {
                    break;
                }
            }
        }
    }

    Ok(())
}
